### ISSUE TITLE
API support for context.Context

### ISSUE THREAD

**Issue by sayrer** [go-mgo/mgo#230]
https://godoc.org/golang.org/x/net/context

It's often desirable to have global cancellation available, regardless of the timeout set for mgo. An API with support for context.Context would solve this.

**Comment by niemeyer** [comment]
The idea is indeed interesting, but let's wait until context goes into the standard library so we don't need to break compatibility once that happens.

**Comment by sayrer** [comment]
is context.Context planned to move into the standard library? if so, wouldn't it make sense to leave this issue open?

**Comment by niemeyer** [comment]
There are conversations about supporting it in some form. I'd prefer to leave this closed for the time being. We support timeouts today, and we're not even sure about the details of standard library support for context.Context. So the answer _today_ to x/net/context would be no. Let's see what happens next and we can discuss again.

**Comment by sayrer** [comment]
Fair enough, but my experience is that the timeouts mgo currently has don't always work.

In particular, a very slow query or a server under load may do enough to keep an mgo socket alive, but still be orders of magnitude too slow to be useful.

**Comment by rgarcia** [comment]
Now that `context` is part of the standard library, should this issue be reopened?

Also, `database/sql` will support `context` in Go 1.8, so I think there will soon be an expectation on other database drivers to support `context`.

**Comment by niemeyer** [comment]
Yes, It's time to fix this. Thanks for raising it.

**Comment by wedneyyuri** [comment]
Hi @niemeyer, can we make pull requests for it?

**Comment by diegobernardes** [comment]
@niemeyer any update on this? Do you need a PR to implement the context?

**Comment by rgarcia** [comment]
FWIW, the official Go driver (in alpha release) has context support https://github.com/mongodb/mongo-go-driver
